### DISCUSSION TITLE
Example of uploading images to server with React.js

### DISCUSSION THREAD

**Discussion by amirhhashemi** [ueberdosis/tiptap#2950] [Showcase]
Uploading images to the server is one of the most challenging things I've done with. I nearly spend one week finding a good way to upload images to the server and show a nice placeholder while it's uploading. It shouldn't be like that. I understand that there are dozens of ways to upload an image to the server and the core team can't promote just one way as the right way, but at least, they should provide some kind of advice, best practices, and examples to do that so people like me don't spend one week to do such a common thing.

Anyway, I created such an example for you :) it's mostly based on the [ProseMirror's image upload example](https://prosemirror.net/examples/upload/). The hard part was that I needed to dive into ProseMirror to understand that, you might too. It is what it is. I think after https://github.com/ueberdosis/tiptap/pull/2919 merged things will get simpler.

codesandbox: https://codesandbox.io/s/react-typescript-forked-c4inxv

## Demo

https://user-images.githubusercontent.com/87268103/177111862-c46b1db4-9989-4387-ad33-24d2b640e697.mp4

For now, you can just select an image from your filesystem. I'm working on adding drag and drop support. Then a placeholder will be shown based on the image's data-url and when the uploading got completed, the placeholder will be 
 replaced with the actual image.

## Known issues
1. Do you see that layout shift when the placeholder gets replaces with the actual image? I don't know why that happens.
2. The image gets inserted after the cursor. it should be inserted before the cursor.

I hope the people that are reading this help me to resolve these issues. I've actually asked for help previously but no one answered.

**Comment by ikorason** [comment]
hey @amirhhashemi did you figure out how to do the drag and drop and with placeholder?

**Reply by bdbch** [reply]
Did you try using https://tiptap.dev/docs/editor/extensions/functionality/filehandler#ondrop?

You could use a Decoration to render a Widget decoration at your drop position showing the upload progress, but it's not something I did yet in that way. What I did prior was using an ImageUpload node that would show progress.
